pkg/controllers/node/kdd.go

```diff
@@ -0,0 +1,204 @@
+// It returns an error if it is determined that there are resources which should be cleaned up, but is unable to do so.
+// It does not return an error if it is successful, or if there is no action to take.
+func (c *NodeController) syncDeleteKDD() error {
+	// Get the backend client.
+	type accessor interface {
+		Backend() bapi.Client
+	}
+	bc := c.calicoClient.(accessor).Backend()
+
+	// Query all IPAM blocks in the cluster, ratelimiting calls.
+	time.Sleep(c.rl.When(RateLimitCalicoList))
+	blocks, err := bc.List(c.ctx, model.BlockListOptions{}, "")
+	if err != nil {
+		return err
+	}
+	c.rl.Forget(RateLimitCalicoList)
+
+	// Build a list of all the nodes in the cluster based on IPAM allocations across all
+	// blocks, plus affinities.
+	nodes := map[string][]model.AllocationAttribute{}
+	for _, kvp := range blocks.KVPairs {
+		b := kvp.Value.(*model.AllocationBlock)
+
+		// Include affinity if it exists. We want to track nodes even
+		// if there are no IPs actually assigned to that node.
+		if b.Affinity != nil {
+			n := strings.TrimLeft(*b.Affinity, "host:")
+			if _, ok := nodes[n]; !ok {
+				nodes[n] = []model.AllocationAttribute{}
+			}
+		}
+
+		// Go through each IPAM allocation, check its attributes for the node it is assigned to.
+		for _, idx := range b.Allocations {
+			if idx == nil {
+				// Not allocated.
+				continue
+			}
+			attr := b.Attributes[*idx]
+
+			// Track nodes based on IP allocations.
+			if val, ok := attr.AttrSecondary["node"]; ok {
+				if _, ok := nodes[val]; !ok {
+					nodes[val] = []model.AllocationAttribute{}
+				}
+
+				// If there is no handle, then skip this IP. We need the handle
+				// in order to release the IP below.
+				if attr.AttrPrimary == nil {
+					log.Warnf("Skipping IP with no handle")
+					continue
+				}
+
+				// Add this allocation to the node, so we can release it later if
+				// we need to.
+				nodes[val] = append(nodes[val], attr)
+			}
+		}
+	}
+	log.Debugf("Nodes in IPAM: %s", nodes)
+
+	// For each node present in IPAM, if it doesn't exist in the Kubernetes API then we
+	// should consider it a candidate for cleanup.
+	for node, allocations := range nodes {
+		// Check if it exists in the Kubernetes API.
+		logc := log.WithField("node", node)
+		if c.nodeExists(node) {
+			logc.Debug("Node still exists, continue")
+			continue
+		}
+
+		// Node exists in IPAM but not in the Kubernetes API. Go through each IP address and
+		// check to see if the pod it references exists. If all the pods on that node are gone,
+		// continue with deletion. If any pod still exists, we skip this node. We want to be
+		// extra sure that the node is gone before we clean it up.
+		canDelete := true
+		for _, a := range allocations {
+			ns := a.AttrSecondary["namespace"]
+			pod := a.AttrSecondary["pod"]
+
+			// TODO: Need to handle IPIP addresses somehow. They are allocated in calico-ipam, but
+			// currently don't have this metadata attached.
+			if ns == "" || pod == "" {
```

Comment:
Should probably have a special attribute that means "This is the IPIP tunnel address for this node"